### PR TITLE
[Feature] Add BigNAS algorithm

### PR DESCRIPTION
**Pull request by wutongshenqiu**
## Motivation

Add BigNAS algorithm.

## Modification

1. Includes modifications mentioned in pr [open-mmlab/mmrazor#215](https://github.com/open-mmlab/mmrazor/pull/215)
2. Pull out the grouping logic of mutator as mixin
3. Add AttentiveNAS and BigNAS search space for mobilenet
4. Add `CalibrateBNMixin` as mixin
5. Add bignas algorithm

## Todo

1. align the training performance.

## Discusstion

1. what kind of backbone is written more clearly. Refer to [BigNASMobileNet](https://github.com/wutongshenqiu/mmrazor/blob/7c2ba5b40a6f3bd3da32412f109e1915e125c50c/mmrazor/models/architectures/backbones/bignas_mobilenet.py) and [AttentiveMobileNet](https://github.com/wutongshenqiu/mmrazor/blob/7c2ba5b40a6f3bd3da32412f109e1915e125c50c/mmrazor/models/architectures/backbones/attentive_mobilenet.py)

## BC-breaking (Optional)

1. Same as pr [open-mmlab/mmrazor#215](https://github.com/open-mmlab/mmrazor/pull/215)

### CODE REVIEW

**Comment by gaoyang07** [comment]
inference accuracy aligned
319cd0f8932028f96306522fbc811e5

File: mmrazor/models/architectures/ops/gml_mobilenet_series.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Dict
```
Comment (sunnyxiaohu) [inline mmrazor/models/architectures/ops/gml_mobilenet_series.py:2]: del this file, if not necessary.

File: mmrazor/structures/subnet/fix_subnet.py [hunk 2]
```diff
@@ -13,9 +13,7 @@ def _dynamic_to_static(model: nn.Module) -> None:
     # Avoid circular import
     from mmrazor.models.architectures.dynamic_ops import DynamicMixin
 
-    def traverse_children(module: nn.Module) -> None:
-        # TODO
-        # dynamicop must have no dynamic child
+    def traverse_children(module: nn.Module, prefix: str = '') -> None:
```
Comment (sunnyxiaohu) [inline mmrazor/structures/subnet/fix_subnet.py:16]: prefix for?

File: mmrazor/models/mutators/channel_mutator/channel_mutator.py [hunk 3]
```diff
@@ -228,7 +229,9 @@ def sample_choices(self) -> Dict[int, Any]:
 
         return random_choices
 
-    def set_choices(self, choices: Dict[int, Any]) -> None:
+    def set_choices(
+            self, choices: Union[Dict[int, Any],
+                                 Callable[[], Dict[Any, Any]]]) -> None:
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutators/channel_mutator/channel_mutator.py:234]: ?

File: mmrazor/models/architectures/dynamic_ops/mixins/dynamic_mixins.py [hunk 4]
```diff
@@ -406,3 +406,97 @@ def to_static_op(self: nn.Linear) -> nn.Module:
...
+            fixed_shape = None
+        return self.static_op_factory(
+            size=fixed_shape,
+            interpolation_type=self._interpolation_type,  # type:ignore
+            align_corners=self._align_corners,  # type:ignore
+            scale_factor=self._scale_factor,  # type:ignore
+            recompute_scale_factor=self._recompute_scale_factor)  # type:ignore
+
+
+class DynamicSequentialMixin(DynamicMixin):
```
Comment (sunnyxiaohu) [inline mmrazor/models/architectures/dynamic_ops/mixins/dynamic_mixins.py:455]: del

**Comment by aptsunny** [comment]
```
11/22 17:53:04 - mmengine - INFO - Epoch(test) [98/98]  min_subnet.accuracy/top1: 75.7480  min_subnet.accuracy/top5: 92.6440
```

gml ->  81.6660 / 77.2580

min子网还差一个点，待check

**Comment by sunnyxiaohu** [comment]
calibrated_sample_nums -> calibrate_sample_num

File: mmrazor/models/algorithms/nas/bignas.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+            self.mutators = built_mutators
+        else:
+            raise TypeError('mutator should be a `dict` but got '
+                            f'{type(mutators)}')
+
+        self.distiller = self._build_distiller(distiller)
+        self.distiller.prepare_from_teacher(self.architecture)
+        self.distiller.prepare_from_student(self.architecture)
+
+        self.strategy = strategy
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/nas/bignas.py:107]: if use stategy, it could be more clear that using `num_samples` to denote the total subnets, e.g. self.num_samples = len(self.strategy_groups[strategy]).

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 6]
```diff
@@ -68,6 +69,7 @@ def __init__(self,
                  num_crossover: int = 25,
                  mutate_prob: float = 0.1,
                  crossover_prob: float = 0.5,
+                 calibrated_sample_nums: int = 4096,
```
[the remark this answers is not in the document]
Comment (aptsunny) [reply mmrazor/engine/runner/evolution_search_loop.py]: Fixed!

File: mmrazor/models/algorithms/nas/bignas.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        for mutator in self.mutators.values():
+            mutator.set_choices(mutator.max_choice)
+
+    def set_min_subnet(self) -> None:
+        """Set min subnet."""
+        for mutator in self.mutators.values():
+            mutator.set_choices(mutator.min_choice)
+
+    @property
+    def search_groups(self) -> Dict:
```
[the remark this answers is not in the document]
Comment (aptsunny) [reply mmrazor/models/algorithms/nas/bignas.py:180]: del

**Review by sunnyxiaohu — APPROVED** [review]
